### PR TITLE
Start to Refactor CassandraStorage

### PR DESCRIPTION
**Pull request by Miles-Garnsey**
This PR begins the refactoring process on CassandraStorage as discussed in #1265. The long term goals are:

1. Split up the monolithic IStorage,  IDistributedStorage interfaces and monolithic CassandraStorage and MemoryStorage classes so that we end up with smaller DAO classes mapping more closely to the types defined in `core`.
2. Ensure return types are aligned with types from `core` where possible and are never `ResultSet`s (which are almost untyped and hard to work with).
3. Bring the codebase into a more manageable state where it is clear which types are collaborating with other types.

This PR is a very small step in that direction, and simply delegates many of the storage operations from CassandraStorage to DAO types. 

Further steps need to be taken to reflect the broad cuts made here in the interfaces and in MemoryStorage, before creating smaller interfaces that the new DAO types adhere to and passing those interfaces (instead of e.g. the larger monolithic IStorage)  to consumers.

### CODE REVIEW

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java [hunk 1]
```diff
@@ -0,0 +1,836 @@
...
+
+    boolean skipMigration = System.getenv().containsKey("REAPER_SKIP_SCHEMA_MIGRATION")
+        ? Boolean.parseBoolean(System.getenv("REAPER_SKIP_SCHEMA_MIGRATION"))
+        : Boolean.FALSE;
+
+    if (skipMigration) {
+      LOG.info("Skipping schema migration as requested.");
+    } else {
+      MigrationManager.initializeAndUpgradeSchema(cassandra, session, config, version, mode);
+    }
```
Comment (adejanovski) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java]:
```suggestion
    version = cassandra.getMetadata().getAllHosts()
        .stream()
        .map(h -> h.getCassandraVersion())
        .min(VersionNumber::compareTo)
        .get();

    boolean skipMigration = System.getenv().containsKey("REAPER_SKIP_SCHEMA_MIGRATION")
        ? Boolean.parseBoolean(System.getenv("REAPER_SKIP_SCHEMA_MIGRATION"))
        : Boolean.FALSE;

    if (skipMigration) {
      LOG.info("Skipping schema migration as requested.");
    } else {
      MigrationManager.initializeAndUpgradeSchema(cassandra, session, config, version, mode);
    }

    this.eventsDao =  new EventsDao(session);
    this.metricsDao  = new MetricsDao(session);
    this.snapshotDao = new SnapshotDao(session);
    this.operationsDao  = new OperationsDao(session);
    this.concurrency = new Concurrency(version, reaperInstanceId, session);
    this.repairUnitDao  = new RepairUnitDao(defaultTimeout, session);
    this.repairSegmentDao = new RepairSegmentDao(concurrency, repairUnitDao, session);
    this.repairScheduleDao = new RepairScheduleDao(repairUnitDao, session);
    this.clusterDao  = new ClusterDao(repairScheduleDao, repairUnitDao, eventsDao, session, objectMapper);
    this.repairRunDao =  new RepairRunDao(repairUnitDao, clusterDao, repairSegmentDao, session);
```

Schema migrations need to happen before creating the DAOs. It looks like their constructors invoke `prepareStatements()` which will do exactly that but on non existing tables.

After moving this block I was able to start Reaper properly.

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/RepairScheduleDao.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+  PreparedStatement insertRepairScheduleByClusterAndKsPrepStmt;
+  PreparedStatement deleteRepairSchedulePrepStmt;
+  PreparedStatement deleteRepairScheduleByClusterAndKsByIdPrepStmt;
+  private final RepairUnitDao repairUnitDao;
+  private final Session session;
+
+
+  public RepairScheduleDao(RepairUnitDao repairUnitDao, Session session) {
+    this.repairUnitDao = repairUnitDao;
+    this.session = session;
```
Comment (adejanovski) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/RepairScheduleDao.java:65]:
```suggestion
    this.session = session;
    prepareStatements()
```

**Review by adejanovski — CHANGES_REQUESTED** [review]
Couple of requested changes to unblock testing.

**Comment by Miles-Garnsey** [comment]
@adejanovski thanks for the input, the tests now pass. Can you please complete your review, and especially take a look at what I've done with the acceptance tests? I've removed all references to Cassandra 2.x to get things running, but maybe we can leave the cases where the cluster to be repaired is Cassandra 2.x, as long as the backend is 3.x+?

I do note that the tests against trunk all seem to be failing, I'm not sure if that is a problem or not given that trunk is presumably often  unstable (or API incompatible with Reaper perhaps? Not sure...)

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -78,27 +78,10 @@ jobs:
     continue-on-error: ${{ matrix.experimental }}
     strategy:
       matrix:
-        cassandra-version: ["binary:1.2.19", "binary:2.0.17", 'binary:2.1.22', 'binary:2.2.19', 'binary:3.0.25', 'binary:3.11.11', 'binary:4.0.1', 'github:apache/trunk']
+        cassandra-version: ['binary:3.0.25', 'binary:3.11.11', 'binary:4.0.1', 'github:apache/trunk']
```
Comment (adejanovski) [inline .github/workflows/ci.yaml]: You'll need to reintroduce the removed entries from the matrix.
This job uses the memory backend and only tests repairs on such versions.

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java [hunk 4]
```diff
@@ -0,0 +1,836 @@
...
+
+  @Override
+  public List<GenericMetric> getMetrics(
+      String clusterName,
+      Optional<String> host,
+      String metricDomain,
+      String metricType,
+      long since) {
+
+    // Compute the hourly buckets since the requested lower bound timestamp
```
Comment (adejanovski) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java]: This comment can be removed.

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java [hunk 5]
```diff
@@ -0,0 +1,836 @@
...
+  public int getSegmentAmountForRepairRunWithState(UUID runId, RepairSegment.State state) {
+    return repairSegmentDao.getSegmentAmountForRepairRunWithState(runId, state);
+  }
+
+  @Override
+  public Collection<RepairRun> getRepairRunsForUnit(UUID repairUnitId) {
+
+    // Grab all ids for the given cluster name
+
+    // Grab repair runs asynchronously for all the ids returned by the index table
```
Comment (adejanovski) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java]: These comments should be removed/moved to the DAO.

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java [hunk 6]
```diff
@@ -0,0 +1,836 @@
...
+  public List<RepairRun> getRepairRunsForClusterPrioritiseRunning(String clusterName, Optional<Integer> limit) {
+    // We've set up the RunState enum so that values are declared in order of "interestingness",
+    // we iterate over the table via the secondary index according to that ordering.
+
+    // Flatten the UUIDs from each status down into a single array.
+    // Merge the two lists and trim.
+
+    // Run an async query on each UUID in the flattened list, against the main repair_run table with
+    // all columns required as an input to `buildRepairRunFromRow`.
+    // Defuture the repair_run rows and build the strongly typed RepairRun objects from the contents.
```
Comment (adejanovski) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java]: These comments should be removed/moved to the DAO.

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java [hunk 7]
```diff
@@ -0,0 +1,836 @@
...
+  @Override
+  public Optional<RepairRun> getRepairRun(UUID id) {
+    return repairRunDao.getRepairRun(id);
+  }
+
+  @Override
+  public Collection<RepairRun> getRepairRunsForCluster(String clusterName, Optional<Integer> limit) {
+
+    // Grab all ids for the given cluster name
+    // Grab repair runs asynchronously for all the ids returned by the index table
```
Comment (adejanovski) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java]: These comment should be removed/moved to the DAO.

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java [hunk 8]
```diff
@@ -0,0 +1,836 @@
...
+  }
+
+  @Override
+  public boolean updateCluster(Cluster newCluster) {
+    return clusterDao.updateCluster(newCluster);
+  }
+
+  private boolean addClusterAssertions(Cluster cluster) {
+
+    // assert we're not overwriting a cluster with the same name but different node list
```
Comment (adejanovski) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java]: This comment should be removed/moved to the DAO.

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java [hunk 9]
```diff
@@ -0,0 +1,836 @@
...
+  }
+
+  public Set<String> getLockedNodesForRun(UUID runId) {
+
+    return concurrency.getLockedNodesForRun(runId);
+  }
+
+  @Override
+  public List<PercentRepairedMetric> getPercentRepairedMetrics(String clusterName, UUID repairScheduleId, Long since) {
+    // Compute the ten minutes buckets since the requested lower bound timestamp
```
Comment (adejanovski) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java]: This comment should be removed/moved to the DAO.

**Review by adejanovski — DISMISSED** [review]
Very nice refactorings!
There are some of the tests in the memory jobs that need to be re-introduced and a few comments that should be deleted.
Otherwise LGTM.
You can merge the PR once the above points have been addressed and CI passes.

**Comment by Miles-Garnsey** [comment]
All changes made as requested, with removal of `dateOf` references to ensure compatibility with `trunk`. I'll merge once/if tests pass.

**Review by adejanovski — APPROVED** [review]

**Pull request merged**